### PR TITLE
Add invalid blocks to blockDAG and keep in justifications to be available for slashing

### PR DESCRIPTION
**Pull request by KentShikama**
## Overview
Provide a brief description of what this PR does, and why it's needed.

### Does this PR relate to an RChain JIRA issue? 
If applicable, add link to corresponding JIRA issue.

https://rchain.atlassian.net/browse/RHOL-467

### Complete this checklist before you submit the PR
- [X] This PR contains no more than 200 lines of code, excluding test code.
- [X] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [X] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 1]
```diff
@@ -277,6 +279,29 @@ object Validate {
     } yield
       maybeCheckPoint match {
         case Some(_) => Right(Valid)
-        case None    => Left(InvalidTransaction)
+        case None    => Right(Valid) // TODO: Fix to Left(InvalidTransaction)
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala]: @birchmd 

```
  // See [[/docs/casper/images/minimal_equivocation_neglect.png]] but cross out genesis block
  it should "not ignore equivocation blocks that are required for parents of proper nodes" in {
    val nodes   = HashSetCasperTestNode.network(validatorKeys.take(3), genesis)
    val deploys = (0 to 5).map(i => ProtoUtil.basicDeploy(i))

    // Creates a pair that constitutes equivocation blocks
    val Some(signedBlock1) = nodes(0).casperEff.deploy(deploys(0)) *> nodes(0).casperEff.createBlock
    val Some(signedBlock1Prime) = nodes(0).casperEff
      .deploy(deploys(1)) *> nodes(0).casperEff.createBlock

    nodes(1).casperEff.addBlock(signedBlock1)
    nodes(0).transportLayerEff.msgQueues(nodes(0).local).clear //nodes(0) misses this block
    nodes(2).transportLayerEff.msgQueues(nodes(2).local).clear //nodes(2) misses this block

    nodes(0).casperEff.addBlock(signedBlock1Prime)
    nodes(1).transportLayerEff.msgQueues(nodes(1).local).clear //nodes(1) misses this block
    nodes(2).receive()
```

For some reason when nodes(2) tries to receive that signedBlock1Prime and it validates the transactions, it says that it doesn't match the tuplespace hash stored in the block. I'm going to sleep on this now but can you think of a reason at quick glance?

**Comment by KentShikama** [comment]
@birchmd PTAL

File: rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala [hunk 2]
```diff
@@ -115,17 +134,7 @@ class ReplayRSpace[C, P, A, R, K](val store: IStore[C, P, A, K], val branch: Bra
...
-                .asScala
-                .collectFirst {
-                  case commRef: COMM =>
-                    runMatcher(Some(commRef)) match {
-                      case Some(x) => Right(x)
-                      case None    => Left(commRef)
-                    }
-                }
-                .get
+              getCommOrDataCandidates(comms.iterator().asScala.toList)
```
Comment (goral09) [inline rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala]:
```scala
comms.toStream.map(x => runMatcher(Some(x))).collectFirst { case Right(x) => x }.getOrElse(comms.head) 
```
should work.

Comment (henrytill) [reply rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala]: (with caveats about `Left` discussed on discord)

We should also try to do the same in `produce`.

**Pull request merged**